manual/tools/caml_tex2.ml

```diff
@@ -248,6 +325,86 @@ type incompatibility =
   | Signature_with_visible_answer of string * int
 exception Incompatible_options of incompatibility
 
+exception Phrase_parsing of string
+
+module Ellipsis = struct
+  (** This module implements the extraction of ellipsis locations
+      from phrases.
+
+      An ellipsis is either an [[@ellipsis]] attribute, or a pair
+      of [[@@@ellipsis.start]...[@@@ellipsis.stop]] attributes. *)
+
+  exception Unmatched_ellipsis of {kind:string; start:int; stop:int}
+  (** raised when an [[@@@ellipsis.start]] or [[@@@ellipsis.stop]] is
+      not paired with another ellipsis attribute *)
+
+  exception Nested_ellipses of {first:int ; second:int }
+  (** raised by [[@@@ellipsis.start][@@@ellipsis.start]] *)
+
+  let extract f x =
+    let transforms = ref [] in
+    let last_loc = ref Location.none in
+    let left_mark = ref None (* stored position of [@@@ellipsis.start]*) in
+    let location _this loc =
+      (* we rely on the fact that the default iterator call first
+         the location subiterator, then the attribute subiterator *)
```

Comment:
On reviewing I didn't think much of this assertion, but upon usage I see bugs caused by the fact that it is wrong. For example the following does not behave correctly:

```
module type DEVICE = sig val draw : picture -> unit[@@ellipsis] end
```

The ellipsis should elide the whole signature item `val draw ...`, but it only elides the `unit` part, because location iterators are called in the wrong order.

The mismatch can be found in the sources of [ast_iterator.ml](https://github.com/ocaml/ocaml/blob/7835cbb/parsing/ast_iterator.ml):

```ocaml

    module_type_declaration =
      (fun this {pmtd_name; pmtd_type; pmtd_attributes; pmtd_loc} ->
         iter_loc this pmtd_name;
         iter_opt (this.module_type this) pmtd_type;
         this.attributes this pmtd_attributes;
         this.location this pmtd_loc
);
```

Instead, in theory we could overload the `attributes` iterator only, and get the location from `this`. I will try to work on a fix along these lines to include in my caml_example PR.

Reply:
In fact I am currently under the impression that there is no nice fix that does not touch `ast_iterator` (which would be the condition to go into 4.07). I started writing the following

```ocaml
let located_attributed_iterator located_attribute =
  let open Ast_iterator in
  let super = default_iterator in
  let rec this = {
    super with
    structure_item = (fun this ({pstr_loc = loc; pstr_desc = desc} as si) ->
      begin match desc with
        | Pstr_eval (_, attrs) -> located_attributes loc attrs
        | Pstr_extension (_, attrs) -> located_attributes loc attrs
      end;
      super.structure_item this si;
      );
```

but the idea of having to extend with definition with all the cases under the sun is killing me, and I don't think this is the right approach.

I would be tempted to revert the ellipsis stuff from 4.07, because I don't like the idea of releasing something that we know is broken. Then for trunk we can discuss either:
- fixing the Ast_iterator control flow (but then I realized that some things have attributes but no locations, so we would probably need to change Parsetree as well if we want to enforce the invariant that all attributable things are located. This is the right thing to do, but it is bound to be long, painful, and break user code.)
- and/or using extension points instead of attributes (that was actually my very first interface idea).

Reply:
@Octachron I'll wait until you confirm that you agree there is no simple/easy fix here before reverting.

Reply:
Reading more carefully `ast_iterator`, I found only 7 exception to the `location; attribute` pattern:
- iter_type_extension
- iter_structure_item
- iter_signature_item
- value_description
- module_declaration
- module_type_declaration
- module_binding
My biased opinion is that writing a custom iterator rewriting those part  would be ugly but acceptable.

Reply:
See [this commit](https://github.com/Octachron/ocaml/commit/3d00d8472ff902a6f8e522bff164789f490a4b12) for some ugliness quantification. 

Concerning your other points, extension nodes were my second idea after pseudo-latex macro; but I found that attribute were the right interface at the user level since there is nothing special going at the code level, and this is really a meta-information. I think it is worth it to try to support this interface.
In term of maintenability,  would it be a bad idea to move (on trunk) the tuned iterator to `parsing/ast_iterator.ml` as another variant of ast iterator?
For the attributable items  that need a more precise location, I noted attribute locations polymorphic variant and object tag, exceptions. This seems like a small set which could be fixed at an ulterior date. 

At least, this my opinion on the spot, I would rethink the issue tomorrow.